### ISSUE TITLE
Problem with namespaces and regex

### ISSUE THREAD

**Issue by ksowa** [socketio/socket.io#4615]
**Describe the bug**
There seems to be a problem with `connection` callback on the server if the following occur:
1. Server uses dynamic namespaces (e.g. `io.of(/^.*$/)...`)
2. Server sends a message to the namespace before any client connects to that namespace
3. When at this point client connects to the server, the `connection` callback is not triggerred 

**To Reproduce**

Socket.IO server version: `4.5.4`

*Server*

```js
import { Server } from "socket.io";

const io = new Server(3000, {
    cors: {
        origin: (origin, callback) => callback(null, origin),
    },
});

io.of(/^.*$/).on('connection', (socket) => {
    console.log('Connected!', socket.id);
});

io.of('testing/namespace').emit('message', 'testing message');
```

Socket.IO client version: `4.5.4`

*Client*

```js
import { io } from "socket.io-client";

const socket = io("ws://localhost:3000/testing/namespace", {});

socket.on("connect", () => {
  console.log(`connect ${socket.id}`);
});
```

**Expected behavior**

The `connection` callback should be triggered regardless if the first message has been emitted to the namespace before or after any clients connect to that namespace.

**Additional context**

If we change the server code to the below, it works without any problem.

(first message emitted after first client connects)

```js
const {Server} = require('socket.io');

const io = new Server(3000, {
    cors: {
        origin: (origin, callback) => callback(null, origin),
    },
});

io.of(/^.*$/).on('connection', (socket) => {
    console.log('Connected!', socket.id);
    io.of('testing/namespace').emit('message', 'test');
});
```

**OR** (namespace is not dynamic):

```js
const {Server} = require('socket.io');

const io = new Server(3000, {
    cors: {
        origin: (origin, callback) => callback(null, origin),
    },
});

io.of('testing/namespace').on('connection', (socket) => {
    console.log('Connected!', socket.id);
});

io.of('testing/namespace').emit('message', 'test');
```

**OR** (client connecting to another namespace):

```js
import { io } from "socket.io-client";

const socket = io("ws://localhost:3000/some/other/namespace", {});

socket.on("connect", () => {
  console.log(`connect ${socket.id}`);
});
```

**Comment by darrachequesne** [comment]
I could indeed reproduce the issue, thanks. This is a known limitation with dynamic namespaces, any existing static namespaces have the priority over the dynamic ones.

We could maybe fix this by running the regex against any existing namespaces:

```js
io.of('test1');

io.of(/^.*$/); // add test1 as a child of the parent namespace (since it matches the regex)

io.of('test2'); // add test2 as a child of the parent namespace (since it matches the regex)
```

That might be a surprising behavior for the end users though. What do you think?

Related: https://github.com/socketio/socket.io/issues/4164

**Comment by darrachequesne** [comment]
@ksowa this should be fixed by https://github.com/socketio/socket.io/commit/0d0a7a22b5ff95f864216c529114b7dd41738d1e, included in version `4.6.1`.

A namespace whose name matches the regex of a parent namespace will now be added as a child of this namespace:

```js
const parentNamespace = io.of(/^\/dynamic-\d+$/);

parentNamespace.on("connection", (socket) => {
  console.log(`connection on namespace ${socket.nsp.name}`);
});
```

And then somewhere else:

```js
io.of("/dynamic-101");
```

Could you please check?

**Comment by ksowa** [comment]
@darrachequesne it is indeed fixed in `4.6.1`! Thank you for the very fast reaction!

**Comment by darrachequesne** [comment]
Awesome, thanks for the feedback :+1:
